### PR TITLE
Improve EntityValueResolver

### PR DESCRIPTION
**Pull request by nicolas-grekas**

### CODE REVIEW

File: src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php [hunk 1]
```diff
@@ -31,24 +31,12 @@
...
-        'auto_mapping' => true,
-        'attribute_only' => false,
-    ];
-
     public function __construct(
         private ManagerRegistry $registry,
-        private ?ExpressionLanguage $language = null,
-        array $defaultOptions = [],
+        private ?ExpressionLanguage $expressionLanguage = null,
+        private MapEntity $defaults = new MapEntity(),
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php:37]: let's use the attribute as the structure that describes options/defaults

File: src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php [hunk 2]
```diff
@@ -61,20 +49,20 @@ public function supports(Request $request, ArgumentMetadata $argument): bool
         }
 
         $options = $this->getOptions($argument);
-        if (null === $options['class']) {
+        if (!$options->class) {
             return false;
         }
 
-        if ($options['attribute_only'] && !$options['has_attribute']) {
+        if (!$this->defaults->autoConvert && !$options->hasAttribute) {
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php]: autoConvert might be clearer, also to ease the migration from FEB

File: src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php [hunk 3]
```diff
@@ -176,59 +162,55 @@ private function getIdentifier(Request $request, array $options, string $name):
...
         }
 
         return false;
     }
 
-    private function findOneBy(string $class, Request $request, array $options): false|object|null
+    private function findOneBy(string $class, Request $request, MapEntity $options): false|object|null
     {
-        if (!$options['mapping']) {
-            if (!$options['auto_mapping']) {
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php:197]: No need for option `auto_mapping`: making `mapping` nullable does the job.

File: src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php [hunk 4]
```diff
@@ -176,59 +162,55 @@ private function getIdentifier(Request $request, array $options, string $name):
...
 
-        if (!$options['mapping']) {
+        if (!$mapping) {
             return false;
         }
 
         // if a specific id has been defined in the options and there is no corresponding attribute
         // return false in order to avoid a fallback to the id which might be of another object
-        if ($options['id'] && null === $request->attributes->get($options['id'])) {
+        if (\is_string($options->id) && null === $request->attributes->get($options->id)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php:192]: `id` can be an array and I suppose we want to rule that out here

File: src/Symfony/Bridge/Doctrine/Attribute/MapEntity.php [hunk 5]
```diff
@@ -17,15 +17,33 @@
...
+    {
+        $clone = clone $this;
+        $clone->class ??= class_exists($class ?? '') ? $class : null;
+        $clone->objectManager ??= $defaults->objectManager;
+        $clone->expr ??= $defaults->expr;
+        $clone->mapping ??= $defaults->mapping;
+        $clone->exclude ??= $defaults->exclude ?? [];
+        $clone->stripNull ??= $defaults->stripNull ?? false;
+        $clone->id ??= $defaults->id;
+        $clone->evictCache ??= $defaults->evictCache ?? false;
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Doctrine/Attribute/MapEntity.php:43]: All options have to be nullable if we want to provide defaults.

**Comment by nicolas-grekas** [comment]
(Please squash-merge)

File: src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php (hunk 2 above)
Comment (jderusse) [reply src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php]: What's about `if (!$this->defaults->autoConvert) {`

- When people use the default option `autoConvert = false` they can add an annotation `#[MapEntity()]` to enable the behavior on desired routes
- When people use the default option `autoConvert = true` they can add an annotation `#[MapEntity(autoConvert: false)]` to disable the behavior on desired routes

This solves the issue reported by @derrabus: When people want to disable the auto-conversion on a few routes, they have to disable the feature on the entire application and add annotation on all the routes.

File: src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php (hunk 3 above)
Comment (jderusse) [reply src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php:197]: good catch

File: src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php]: Good idea, updated, and renamed to `map`.

File: src/Symfony/Bridge/Doctrine/Tests/ArgumentResolver/EntityValueResolverTest.php [hunk 6]
```diff
@@ -85,7 +85,7 @@ public function testSupportWithoutClass()
     public function testSupportWithoutAttribute()
     {
         $registry = $this->getMockBuilder(ManagerRegistry::class)->getMock();
-        $converter = new EntityValueResolver($registry, null, ['attribute_only' => true]);
+        $converter = new EntityValueResolver($registry, null, new MapEntity(map: false));
```
Comment (jderusse) [inline src/Symfony/Bridge/Doctrine/Tests/ArgumentResolver/EntityValueResolverTest.php]:
```suggestion
        $converter = new EntityValueResolver($registry, null, new MapEntity(disabled: true));
```

**Pull request merged**